crates/api_common/src/utils.rs

```diff
@@ -954,31 +963,44 @@ pub async fn purge_user_account(person_id: PersonId, context: &LemmyContext) ->
   Ok(())
 }
 
-pub enum EndpointType {
-  Community,
-  Person,
-  Post,
-  Comment,
-  PrivateMessage,
-}
-
-/// Generates an apub endpoint for a given domain, IE xyz.tld
-pub fn generate_local_apub_endpoint(
-  endpoint_type: EndpointType,
-  name: &str,
-  domain: &str,
-) -> Result<DbUrl, ParseError> {
-  let point = match endpoint_type {
-    EndpointType::Community => "c",
-    EndpointType::Person => "u",
-    EndpointType::Post => "post",
-    EndpointType::Comment => "comment",
-    EndpointType::PrivateMessage => "private_message",
+pub enum ObjectType {
+  Community(String),
+  Person(String),
+  Post(PostId),
+  Comment(CommentId),
+  PrivateMessage(PrivateMessageId),
+}
+
+/// Generates url for object type on local domain
+pub fn local_url(endpoint_type: impl Into<ObjectType>, settings: &Settings) -> LemmyResult<DbUrl> {
+  let (point, name) = match endpoint_type.into() {
+    ObjectType::Community(name) => ("c", name),
+    ObjectType::Person(name) => ("u", name),
+    ObjectType::Post(id) => ("post", id.to_string()),
+    ObjectType::Comment(id) => ("comment", id.to_string()),
+    ObjectType::PrivateMessage(id) => ("private_message", id.to_string()),
   };
 
+  let domain = settings.get_protocol_and_hostname();
```

Comment:
Since you're only ever using this protocol and hostname from the settings, I think it'd be *much* cleaner to just add a trait that each of these implement.

crates/db_schema/src/traits.rs

```rust
pub trait ToLocalUrl {
  fn local_url(&self, protocol_and_hostname: &str) -> String;
}
```

crates/db_schema/src/source/comment.rs

```rust
impl ToLocalUrl for Comment {
  fn local_url( ...
```

Reply:
Taking self doesnt work because in some places this is called before an object is constructed, eg `crates/api_crud/src/user/create.rs`

Reply:
Since Community and Person are the outliers (since they use a string and need to be generated beforehand), then you can just not use the trait for those ones (or make a different trait for the actor types), and have a `local_url(name: &str, protocol_and_hostname: &str)`

Ideally they'd work the same way that the others do, where the ap_id column is filled by triggers, but that could be done some other time. cc @dullbananas

Reply:
Changed it to separate methods in the db structs, its actually much better this way. But using traits wouldnt give any benefit here as its not used generically (there are some other useless traits in the codebase too I believe).